### PR TITLE
Docs/git strategies

### PR DESCRIPTION
**Pull request by defaye**
Add documentation outlining Git best practices and strategies for effective usage. This includes creating a new branch for each feature or bug fix, using descriptive commit messages, using atomic commits, regularly pushing changes to a remote repository, using pull requests for code review and collaboration, reviewing and merging pull requests promptly, and using code review tools or automated tests. Additionally, include strategies for effective Git usage, such as writing good commit messages, creating small, atomic commits, using Git's patch option to pick out atomic commits, treating local commits as mutable, using rebase to restructure commits before pushing, and searching through revision histories.

### CODE REVIEW

File: 03-project-delivery/04-tools.md [hunk 1]
```diff
@@ -1,5 +1,85 @@
...
+
+Use descriptive commit messages that ideally follow the [Commitlint](https://commitlint.js.org/) style, which uses a standardised format for commit messages that includes a type, scope, and subject. For example, a commit message might start with "docs: update README with new information" or "feat: add new feature to user profile page".
+
+#### Use atomic commits
+
+Use atomic commits that are small and focused on a single logical change. An atomic commit is a self-contained unit of work that can be applied or reverted independently of other changes.
+
+#### Regularly push changes to a remote repository
+
+Regularly push changes to a remote repository (e.g., GitHub) to ensure that your work is backed up and accessible to others on the team. This also helps avoid loss of work due to hardware failure or other issues.
```
Comment (jalada) [inline 03-project-delivery/04-tools.md:25]:
```suggestion
Regularly push changes to GitHub to ensure that your work is backed up and accessible to others on the team. This also helps avoid loss of work due to hardware failure or other issues.
```

File: 03-project-delivery/04-tools.md [hunk 2]
```diff
@@ -1,5 +1,85 @@
...
+
+Create a new branch for each feature or bug fix, and merge it back into the main branch only after testing and review. This helps keep the main branch stable and avoids conflicts with other developers' work.
+
+#### Use descriptive commit messages
+
+Use descriptive commit messages that ideally follow the [Commitlint](https://commitlint.js.org/) style, which uses a standardised format for commit messages that includes a type, scope, and subject. For example, a commit message might start with "docs: update README with new information" or "feat: add new feature to user profile page".
+
+#### Use atomic commits
+
+Use atomic commits that are small and focused on a single logical change. An atomic commit is a self-contained unit of work that can be applied or reverted independently of other changes.
```
Comment (jalada) [inline 03-project-delivery/04-tools.md:21]: Do we really think every commit should be revertible independently? Is that really a thing?

File: 03-project-delivery/04-tools.md [hunk 3]
```diff
@@ -1,5 +1,85 @@
...
+
+#### Use code review tools or automated tests
+
+Use code review tools or automated tests to check for coding standards, performance, security, and other quality criteria. This helps ensure that the code is of high quality and meets our standards for maintainability and reliability.
+
+By following these guidelines, we can work together effectively and maintain a high standard of quality for our codebase.
+
+## Strategies for Effective Git Usage
+
+A good revision history is essential for software development. Maintainability is a critical aspect of any software development project, and one of the most critical factors in maintainability is the revision history. The revision history captures the changes made to the codebase over time, and it's a valuable resource for understanding why the code is the way it is. Unfortunately, many teams don't put enough effort into maintaining their revision history, leading to lost institutional knowledge and decreased maintainability.
```
Comment (jalada) [inline 03-project-delivery/04-tools.md:43]: I dunno if this paragraph makes sense in a handbook (did ChatGPT write this?)

File: 03-project-delivery/04-tools.md [hunk 4]
```diff
@@ -1,5 +1,85 @@
...
+
+### Write Good Commit Messages
```
Comment (jalada) [inline 03-project-delivery/04-tools.md:45]: I dislike title case btw, can we un-titlecase these? I don't think the handbook is title-cased elsewhere

File: 03-project-delivery/04-tools.md [hunk 5]
```diff
@@ -1,5 +1,85 @@
...
+
+A good commit message should capture both what the change does and why it was made. It should also provide enough context so that anyone reading the message can understand the change. For example, instead of writing "Fix bug," write "Fix bug causing crash when user enters invalid input."
```
Comment (jalada) [inline 03-project-delivery/04-tools.md:47]: This feels like a duplicate of the standard above.

File: 03-project-delivery/04-tools.md [hunk 6]
```diff
@@ -1,5 +1,85 @@
...
+
+### Create Small, Atomic Commits
+
+Commits should focus on doing one thing, making them easier to review and understand. For example, instead of combining refactoring and feature work into one commit, separate them into two separate commits.
```
Comment (jalada) [inline 03-project-delivery/04-tools.md:51]: Didn't we just have this?

File: 03-project-delivery/04-tools.md [hunk 7]
```diff
@@ -1,5 +1,85 @@
...
+* Use `git add -p` to interactively stage changes to a file and create atomic commits, making it easier to track changes and simplify the review process.
+* Use the options `y` to stage a hunk of code, `n` to skip it, and `s` to split it into smaller hunks.
+* Use `q` to quit the interactive mode and return to the command prompt.
+
+### Treat Local Commits as Mutable
+
+Until a commit is merged into the main branch, it can be modified or even deleted. Use the amend option to modify the most recent commit or create fix-up commits and squash them down with an auto-squash during a rebase.
+
+* Use git `commit --amend` to edit the message of a specific commit and `commit --amend --no-edit` if you don't need to edit the message.
+* Use the force wisely: Use `git push --force-with-lease` to force push changes to a remote branch while ensuring that you don't overwrite any changes made by others since your last pull.
```
Comment (jalada) [inline 03-project-delivery/04-tools.md:66]: I don't know if this is useful for (say) a junior to read.

File: 03-project-delivery/04-tools.md [hunk 8]
```diff
@@ -1,5 +1,85 @@
...
+  * Use this option carefully and only when it makes sense to rewrite the commit history, such as when tidying up a commit history prior to submitting a pull request for review.
+
+### Use Rebase to Restructure Commits Before Pushing
+
+Use `git rebase -i <commit>` e.g. `git rebase -i HEAD~5` to interactively rebase the commit history and restructure commits, such as changing commit messages, squashing commits, and reordering commits.
```
Comment (jalada) [inline 03-project-delivery/04-tools.md:71]: This probably needs an explanation of what the goal is, e.g. imagine a junior reading it.

**Review by jalada — CHANGES_REQUESTED** [review]
Feedback from a quick pass.

File: 03-project-delivery/04-tools.md (hunk 3 above)
Comment (defaye) [reply 03-project-delivery/04-tools.md:43]: Yep. This is a inspired by some lessons learned from [A Branch in Time](https://www.youtube.com/watch?v=1NoNTqank_U) (via https://tekin.co.uk/2019/02/a-talk-about-revision-histories)

I wanted to convey the importance of Git commits as an important historical document of a project. Though maybe I'm not understanding what should and shouldn't become content for the handbook 🤔

File: 03-project-delivery/04-tools.md [hunk 9]
```diff
@@ -1,5 +1,66 @@
...
+
+* Use git `commit --amend` to edit the message of a specific commit and `commit --amend --no-edit` if you don't need to edit the message.
+
+### Use rebase to restructure commits before pushing
+
+Let's say you have a feature branch ready to push. You check the Git log and see that you have some poorly written commits, a reverted change, and two commits that straddle another that ought to have been committed together. You're in rebase territory.
+
+Use `git rebase -i <commit>` (e.g. `git rebase -i HEAD~5` to rebase the last 5 commits) to interactively rebase the commit history and restructure commits, such as changing commit messages, squashing commits, and reordering commits.
+Use the options `pick` to keep a commit as-is, `squash` to combine a commit with the previous commit (up), and `edit` to stop at a specific commit for further changes. You can cut and move lines to re-order the sequence of commits too.
+* Don't be afraid to try rebase. You can always use `git rebase --abort` to cancel it if you get unstuck.
```
Comment (JonathanAndrews) [inline 03-project-delivery/04-tools.md:55]: Hmmm... how about

```suggestion
* Don't be afraid to try rebase. You can always use `git rebase --abort` to cancel it if you come unstuck.
```

or 
```suggestion
* Don't be afraid to try rebase. You can always use `git rebase --abort` to cancel it if you get yourself in a pickle.
```
?

File: 03-project-delivery/04-tools.md [hunk 10]
```diff
@@ -1,5 +1,66 @@
...
+
+To use Git effectively, we follow these best practices:
+
+#### Create a new branch for each feature or bug fix
+
+Create a new branch for each feature or bug fix, and merge it back into the main branch only after testing and review. This helps keep the main branch stable and avoids conflicts with other developers' work.
+
+#### Use descriptive commit messages
+
+We're not super prescriptive on commit message format but if you want to follow a standardised format, you could use the [Commitlint](https://commitlint.js.org/) style, which uses a standardised format for commit messages that includes a type, scope, and subject. For example, a commit message might start with "docs: update README with new information" or "feat: add new feature to user profile page".
```
Comment (JonathanAndrews) [inline 03-project-delivery/04-tools.md:17]: Meta 🤘
 Screenshot 2023-04-06 at 08 39 59

**Review by JonathanAndrews — APPROVED** [review]
Look's good to me!
One copy change 👍

**Pull request merged**